Automattic/woocommerce-payments
# Allow WooPay to request session data from merchant

#### Changes proposed in this Pull Request


This PR is to address a performance optimization in the WooPay Direct Checkout feature. This PR is one half of the fix for https://github.com/Automattic/woopay/issues/2480. This PR should be tested in tandem with https://github.com/Automattic/woopay/pull/2529.



The overall feature is to send a minimum amount of session data to WooPay to start, then let WooPay request the rest of the session data only if it is needed (i.e. the shopper is already logged into WooPay). If the shopper is not logged into WooPay the full session data should not be requested and the shopper should be taken to the merchant checkout page. More details about the feature and this optimization can be found in https://github.com/Automattic/woopay/issues/2480.

To accomplish that goal, this PR adds the following:
- The ability to retrieve the minimum session data either on page load or through AJAX if necessary.
- ~~Pass the minimum session data to the connect page and receive a redirect URL.~~
- Pass the encrypted and signed minimum session data to WooPay with a redirect.
- A new REST endpoint that allows WooPay to request the full session data from the merchant.

The points of failure I focused on in this PR:
- Inability to get the minimum session data.
- Inability to retrieve full session data from the merchant.

In all cases when there is a failure there, we should default to taking the shopper to the merchant checkout page.

**Notes**
One key thing to note is the need to unset the `preloaded_requests` on the full session data response. This is unfortunate as it reduces some of the performance benefits. The reason this had to be unset is that the preloaded requests use `rest_preload_api_request` which just replicates a REST request but with no headers, which means no `Cart-Token`. This leads the cart not being found and instead returning an empty cart object breaking things on WooPay. I experimented with other ways to request the cart data doing something like this:

```
add_filter( 'woocommerce_store_api_disable_nonce_check', '__return_true' );
$cart_request = new WP_REST_Request('GET', '/wc/store/v1/cart' );
$cart_request->set_headers( $woopay_request->get_headers() );
$cart_data = rest_do_request( $cart_request );

$checkout_request = new WP_REST_Request('GET', '/wc/store/v1/checkout' );
$checkout_request->set_headers( $woopay_request->get_headers() );
$checkout_data = rest_do_request( $checkout_request );
remove_filter( 'woocommerce_store_api_disable_nonce_check', '__return_true' );
```

But ran into the problem that the initial requests had been fired and so WooCommerce wasn't trying to load the cart again since it already loaded it on the first requests in `get_init_session_request` and it generally seemed "too late" to update the current session to use the Cart Token. **Given the time constraints I opted to leave the unset in for now and revisit in a separate issue, but if we find a good solution I'd love to add it to this PR!**





#### Testing instructions

All tests should start with WooPay enabled, the `_wcpay_feature_woopay_direct_checkout` option set to 1 (enabled), and an item in the cart.

- When the shopper is logged out of WooPay, clicking the `Proceed to Checkout` button should take the shopper to the merchant checkout. If third party cookies are disabled, there will be a request to `woopay/?checkout_redirect=1&blog_id=...` that results in a 302 redirect to the merchant checkout. In this case nothing should ever get stored on WooPay for the session.
- When the shopper is logged in to WooPay,  clicking the `Proceed to Checkout` button should take the shopper to the WooPay checkout. If third party cookies are disabled, there will be a request to `woopay/?checkout_redirect=1&blog_id=...` that results in a 302 redirect to WooPay with a `platform_checkout_key`. 
- Test the existing flows are unaffected:
    - Using the WooPay Express checkout button on the cart and checkout page.
    - Using the OTP modal on checkout page after entering an email address.
    - Checking out without WooPay.
- Test in both blocks and shortcode cart.
- Test with a shopper that is logged into the merchant store.
- Test with the WooPay PR running on the staging sandbox to confirm environment compatibility. 

-------------------


**Post merge**
